backends/p4tools/common/compiler/midend.h

```diff
@@ -64,7 +64,11 @@ class MidEnd : public PassManager {
 
     /// Add the list of default passes to the mid end. This is not part of the initializer because
     /// some targets may add their own passes to the beginning of the pass list.
-    void addDefaultPasses();
+    virtual void addDefaultPasses();
+
+    /// Add passes that break type checking. These passes may involve  IR modifications the front
+    /// end type checker does not recognize.
+    virtual void addNonTypeCheckingPasses();
```

Comment:
What is the problem here? It seems very suspicious to have IR that does not type check. If this is because of custom IR nodes, it might be better to use a derived type check that supports them. Or at least (for now?) comment more on the reasons why this does not type check.

Reply:
All the passes contained in this list perform transformations that are useful for the interpreter but TypeChecking can fail on.

The two most problematic ones is the `HSIndexSimplifier` which converts array indices to members and `ConvertVarbits` which creates a new varbit type that is not recognized.

`CastBooleanTableKeys` converts boolean table keys to bit<1>, which type checking also might not support (I have not checked). 

All these passes could be handled in the interpreter but it might be a fair bit of work.